### PR TITLE
test: Additional regression scenarios for gsheet

### PR DESCRIPTION
**Pull request by sarojsarab**
## Description
- Added following  regression scenarios for gsheet
   1. Create query from active ds tab
   2. Create query from ds details page
   3. Generate CRUD page from active ds
   4. Generate CRUD page from entity explorer
   5. Added test to add query from global search

#### Type of change
- Cypress
#### How Has This Been Tested?
- [ ] Cypress
#### QA activity:
- [ ] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#speedbreakers-) have been covered
- [ ] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#areas-of-interest-)
- [ ] Test plan has been peer reviewed by project stakeholders and other QA members
- [ ] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [ ] Cypress test cases have been added and approved by SDET/manual QA
- [ ] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

### CODE REVIEW

File: app/client/cypress/e2e/GSheet/Misc_Spec.ts [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    // Click on generate page button and verify the page is generated
+    agHelper.GetNClick(dataSources._generatePageBtn);
+    assertHelper.AssertNetworkStatus("@replaceLayoutWithCRUDPage", 201);
+    agHelper.AssertContains("Successfully generated a page");
+    assertHelper.AssertNetworkStatus("@getActions", 200);
+    assertHelper.AssertNetworkStatus("@postExecute", 200);
+    assertHelper.AssertNetworkStatus("@updateLayout", 200);
+    agHelper.GetNClick(dataSources._visibleTextSpan("Got it"));
+
+    // Verify the table data
```
Comment (Aishwarya-U-R) [inline app/client/cypress/e2e/GSheet/Misc_Spec.ts:137]: DeployApp & test below.

File: app/client/cypress/support/Pages/DataSources.ts [hunk 2]
```diff
@@ -892,7 +898,7 @@ export class DataSources {
     if (toNavigateToActive) this.NavigateToActiveTab();
     cy.get(this._datasourceCard, { withinSubject: null })
       .find(this._activeDS)
-      .contains(datasourceName)
+      .contains(new RegExp("^" + datasourceName + "$", "g")) //This regex is to exact match the datasource name
```
Comment (Aishwarya-U-R) [inline app/client/cypress/support/Pages/DataSources.ts:901]: Thanks for doing this.. This has been on my list to correct for so long.

File: app/client/cypress/e2e/GSheet/Misc_Spec.ts (hunk 1 above)
Comment (sarojsarab) [reply app/client/cypress/e2e/GSheet/Misc_Spec.ts]: Done!

File: app/client/cypress/e2e/GSheet/Misc_Spec.ts [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    const data = GSHEET_DATA.filter((item) => item.rowIndex === "0")[0];
```
[the remark this answers is not in the document]
Comment (sarojsarab) [reply app/client/cypress/e2e/GSheet/Misc_Spec.ts:100]: Done!

File: app/client/cypress/support/Pages/DataSources.ts (hunk 2 above)
Comment (sarojsarab) [reply app/client/cypress/support/Pages/DataSources.ts:901]: 👍

File: app/client/cypress/e2e/GSheet/Misc_Spec.ts [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+  });
+
+  it("2. Add query from edit datasource page and verify", () => {
+    dataSources.CreateQueryForDS(dataSourceName, "", "fetch_many", false);
+    dataSources.ValidateNSelectDropdown("Operation", "Fetch Many");
+    dataSources.ValidateNSelectDropdown("Entity", "Sheet Row(s)");
+    agHelper.Sleep(500);
+    dataSources.ValidateNSelectDropdown("Spreadsheet", "", spreadSheetName);
+    dataSources.ValidateNSelectDropdown("Sheet name", "", "Sheet1");
+    dataSources.RunQueryNVerifyResponseViews(GSHEET_DATA.length);
```
Comment (Aishwarya-U-R) [inline app/client/cypress/e2e/GSheet/Misc_Spec.ts:70]: Also verify headers using  `dataSources.AssertQueryResponseHeaders(columnHeaders);`

File: app/client/cypress/e2e/GSheet/Misc_Spec.ts [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+    assertHelper.AssertNetworkStatus("@replaceLayoutWithCRUDPage", 201);
+    agHelper.AssertContains("Successfully generated a page");
+    assertHelper.AssertNetworkStatus("@getActions", 200);
+    assertHelper.AssertNetworkStatus("@postExecute", 200);
+    assertHelper.AssertNetworkStatus("@updateLayout", 200);
+    agHelper.GetNClick(dataSources._visibleTextSpan("Got it"));
+    assertHelper.AssertNetworkStatus("@updateLayout", 200);
+
+    //deploy the app and verify the table data
+    deployMode.DeployApp(locators._widgetInDeployed("tablewidget"));
```
Comment (Aishwarya-U-R) [inline app/client/cypress/e2e/GSheet/Misc_Spec.ts:99]: Use -     `deployMode.DeployApp(locators._widgetInDeployed(draggableWidgets.TABLE));`

File: app/client/cypress/e2e/GSheet/Misc_Spec.ts [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+    });
+    table.ReadTableRowColumnData(0, 3, "v1").then((cellData) => {
+      expect(cellData).to.eq(data.specialChars);
+    });
+    table.ReadTableRowColumnData(0, 4, "v1").then((cellData) => {
+      expect(cellData).to.eq(data.product_name);
+    });
+
+    //Validating loaded JSON form
+    cy.xpath(locators._spanButton("Update")).then((selector) => {
```
Comment (Aishwarya-U-R) [inline app/client/cypress/e2e/GSheet/Misc_Spec.ts:118]: Use `agHelper.GetElement`

File: app/client/cypress/e2e/GSheet/Misc_Spec.ts [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+    gsheetHelper.DeleteSpreadsheetQuery(dataSourceName, spreadSheetName);
+    cy.get("@postExecute").then((interception: any) => {
+      expect(interception.response.body.data.body.message).to.deep.equal(
+        "Deleted spreadsheet successfully!",
+      );
+    });
+    homePage.NavigateToHome();
+    homePage.DeleteApplication(appName);
+    deployMode.NavigateBacktoEditor();
+    table.WaitUntilTableLoad();
```
Comment (Aishwarya-U-R) [inline app/client/cypress/e2e/GSheet/Misc_Spec.ts:199]: Why are these 2 stmts needed?

**Review by Aishwarya-U-R — COMMENTED** [review]
@sarojsarab - Also please run the new & existing GSheet tests on this PR along with normal regression cases.

File: app/client/cypress/e2e/GSheet/Misc_Spec.ts (hunk 7 above)
Comment (sarojsarab) [reply app/client/cypress/e2e/GSheet/Misc_Spec.ts:199]: no this is not required here, it was added by mistake.

File: app/client/cypress/e2e/GSheet/Misc_Spec.ts (hunk 6 above)
Comment (sarojsarab) [reply app/client/cypress/e2e/GSheet/Misc_Spec.ts]: done|

File: app/client/cypress/e2e/GSheet/Misc_Spec.ts (hunk 4 above)
Comment (sarojsarab) [reply app/client/cypress/e2e/GSheet/Misc_Spec.ts:81]: Thanks for suggesting this, fixed.

File: app/client/cypress/e2e/GSheet/Misc_Spec.ts (hunk 5 above)
Comment (sarojsarab) [reply app/client/cypress/e2e/GSheet/Misc_Spec.ts]: Done!

**Comment by sarojsarab** [comment]
I have run the gsheet workflow, below are the details
Workflow run: https://github.com/appsmithorg/appsmith/actions/runs/5723475022
Cypress Dashboard: [Click Here!](https://cloud.cypress.io/projects/eyxvp8/runs/38095)

File: app/client/cypress/support/Pages/DataSources.ts [hunk 8]
```diff
@@ -1761,4 +1762,17 @@ export class DataSources {
...
+
+  public AddQueryFromGlobalSearch(datasourceName: string) {
+    this.agHelper.GetNClick(this._globalSearchTrigger);
+    this.agHelper.Sleep(500);
+    this.agHelper.TypeText(this._globalSearchInput, datasourceName);
+    this.agHelper.Sleep(500);
+    this.agHelper
+      .GetElement(this._globalSearchOptions)
+      .contains(new RegExp("^" + datasourceName + "$", "g"))
+      .click({ force: true });
```
Comment (Aishwarya-U-R) [inline app/client/cypress/support/Pages/DataSources.ts:1774]: can `agHelper.GetNClickByContains` be used?

**Comment by Aishwarya-U-R** [comment]
> Click Here!

Ah, why does it take [~43 mins](https://cloud.cypress.io/projects/eyxvp8/runs/38095/test-results?actions=%5B%5D&browsers=%5B%5D&groups=%5B%5D&isFlaky=%5B%5D&modificationDateRange=%7B%22startDate%22%3A%221970-01-01%22%2C%22endDate%22%3A%222038-01-19%22%7D&orderBy=EXECUTION_ORDER&oses=%5B%5D&specs=%5B%5D&statuses=%5B%5D&testingTypesEnum=%5B%5D) only for Gsheet tests?

File: app/client/cypress/support/Pages/DataSources.ts (hunk 8 above)
Comment (sarojsarab) [reply app/client/cypress/support/Pages/DataSources.ts]: Updated to use the agHelper function.

**Comment by sarojsarab** [comment]
> > Click Here!
> 
> Ah, why does it take [~43 mins](https://cloud.cypress.io/projects/eyxvp8/runs/38095/test-results?actions=%5B%5D&browsers=%5B%5D&groups=%5B%5D&isFlaky=%5B%5D&modificationDateRange=%7B%22startDate%22%3A%221970-01-01%22%2C%22endDate%22%3A%222038-01-19%22%7D&orderBy=EXECUTION_ORDER&oses=%5B%5D&specs=%5B%5D&statuses=%5B%5D&testingTypesEnum=%5B%5D) only for Gsheet tests?

As we are depending on 3rd party api (google api) some time it takes time to respond, and it increases the run time also. Cypress also sometime take lot of time to execute the tests, it takes 4 mins for misc_spec.ts but in ci it takes more than 5 mins

**Review by Aishwarya-U-R — APPROVED** [review]
Approving Cypress changes.

**Comment by Aishwarya-U-R** [comment]
@sarojsarab - Merged this now, please trigger new run for GSheets.

**Pull request merged**